Goto-harness support to selectively havoc struct members

xbauch opened the pull request:
By means of a new option `havoc-member` the user can now specify a struct-within-struct path, e.g. `struct_outer.struct_inner.data` to specify only which member they want to havoc.

Includes a test.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d9a1bb4683ff99596f0c482608c98f1d19fe2d62).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/136301109

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f096f4dd0be170b5721dd8329cbe30db38dec198).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/136317859

hannes-steffenhagen-diffblue commented on src/goto-harness/common_harness_generator_options.h:
```diff
@@ -16,6 +16,7 @@ Author: Diffblue Ltd.
 #define COMMON_HARNESS_GENERATOR_MAX_ARRAY_SIZE_OPT "max-array-size"
 #define COMMON_HARNESS_GENERATOR_FUNCTION_POINTER_CAN_BE_NULL_OPT              \
   "function-pointer-can-be-null"
+#define COMMON_HARNESS_GENERATOR_MEMBER_SELECTION_OPT "member-selection"
```
I don’t really like the name `member-selection`, following the pattern it should be called something like `nondet-member` or `havoc-member`

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -85,6 +86,22 @@ bool recursive_initialization_configt::handle_option(
       [](const std::string &opt) -> irep_idt { return irep_idt{opt}; });
     return true;
   }
+  else if(option == COMMON_HARNESS_GENERATOR_MEMBER_SELECTION_OPT)
```
Correct me if I’m wrong, but this looks as if you can only have one member havoc'd at a time? Because that shouldn’t be the case.

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented:
Linter failures should be addressed with `// NOLINT` to explicitly supress linting.

petr-bauch committed:
Add support to selectively havoc struct members

to goto-harness. By means of a new option `member-selection` the user can now
specify a struct-within-struct path, e.g. `struct_outer.struct_inner.data` to
specify only which member they want to havoc. Includes a test.

petr-bauch committed:
Extend to support multiple members for havocing

via a comma-separated list. Plus one more test.

NlightNFotis reviewed (approved):
LGTM

hannes-steffenhagen-diffblue reviewed (approved):
:+1:

hannes-steffenhagen-diffblue commented:
Now we’d probably also want things like “these two fields in this struct can point to the same variable” or “this field in that struct is the length of the array in that struct”, but not in this PR
